File: modules/desktop-client-sdk/src/main/java/com/avandocmsg/messenger/desktop/sdk/attachments/AttachmentPathResolver.java
package com.avandocmsg.messenger.desktop.sdk.attachments;

import java.nio.file.Path;
import java.text.Normalizer;
import java.time.LocalDate;

public final class AttachmentPathResolver {

    private final Path downloadsRoot;
    private final String profileSlug;

    public AttachmentPathResolver(Path downloadsRoot, String profileSlug) {
        this.downloadsRoot = downloadsRoot;
        this.profileSlug = profileSlug;
    }

    public Path resolve(String serverSlug, String fileId, String originalName) {
        return resolve(serverSlug, fileId, originalName, LocalDate.now());
    }

    public Path resolve(String serverSlug, String fileId, String originalName, LocalDate at) {
        var safeServer = slug(serverSlug);
        var safeName = slug(originalName);
        if (safeName.isBlank()) {
            safeName = "file";
        }
        return downloadsRoot
            .resolve("KorusMessenger")
            .resolve(slug(profileSlug))
            .resolve("attachments")
            .resolve(safeServer)
            .resolve(String.valueOf(at.getYear()))
            .resolve(String.format("%02d", at.getMonthValue()))
            .resolve(fileId + "-" + safeName);
    }

    private static String slug(String input) {
        var n = Normalizer.normalize(input.trim(), Normalizer.Form.NFD)
            .replaceAll("\\p{M}+", "")
            .replaceAll("[^a-zA-Z0-9._-]+", "_");
        if (n.length() > 64) {
            n = n.substring(0, 64);
        }
        return n.isBlank() ? "item" : n;
    }
}

File: mobile/mobile-client-sdk/src/commonMain/kotlin/com/avandocmsg/messenger/mobile/sdk/attachments/AttachmentPathResolver.kt
package com.avandocmsg.messenger.mobile.sdk.attachments

import java.nio.file.Path
import java.time.LocalDate
import java.text.Normalizer

class AttachmentPathResolver(
    private val downloadsRoot: Path,
    private val profileSlug: String
) {
    fun resolve(serverSlug: String, fileId: String, originalName: String, at: LocalDate = LocalDate.now()): Path {
        val safeServer = slug(serverSlug)
        val safeName = slug(originalName).ifBlank { "file" }
        val dir = downloadsRoot
            .resolve("KorusMessenger")
            .resolve(slug(profileSlug))
            .resolve("attachments")
            .resolve(safeServer)
            .resolve(at.year.toString())
            .resolve(at.monthValue.toString().padStart(2, '0'))
        return dir.resolve("$fileId-$safeName")
    }

    private fun slug(input: String): String {
        val n = Normalizer.normalize(input.trim(), Normalizer.Form.NFD)
            .replace(Regex("\\p{M}+"), "")
            .replace(Regex("[^a-zA-Z0-9._-]+"), "_")
            .take(64)
        return n.ifBlank { "item" }
    }
}
